cpp/src/slice2py/PythonUtil.cpp

```diff
@@ -2520,122 +2604,139 @@ Slice::Python::CodeVisitor::writeDocstring(const OperationPtr& op, DocstringMode
+        if (!comment.description.empty())
+        {
+            _out << nl;
+        }
+
+        _out << nl << "Parameters";
+        _out << nl << "----------";
+        for (const auto& param : inParams)
         {
-            string fixed = fixIdent(*q);
-            _out << nl << fixed << " -- ";
-            StringMap::const_iterator r = comment.params.find(*q);
+            string fixed = fixIdent(param->name());
+            _out << nl << fixed << " : " << typeToDocstring(param->type(), param->optional());
+            StringMap::const_iterator r = comment.params.find(param->name());
             if (r == comment.params.end())
             {
                 r = comment.params.find(fixed); // Just in case.
             }
             if (r != comment.params.end())
             {
-                _out << r->second;
+                _out << nl << "    " << r->second;
             }
         }
-        if (mode == DocAsyncBegin)
-        {
-            _out << nl << "_response -- The asynchronous response callback." << nl
-                 << "_ex -- The asynchronous exception callback." << nl << "_sent -- The asynchronous sent callback.";
-        }
-        if (mode == DocSync || mode == DocAsync || mode == DocAsyncBegin)
+        if (mode == DocSync || mode == DocAsync)
         {
             const string contextParamName = getEscapedParamName(op, "context");
-            _out << nl << contextParamName << " -- The request context for the invocation.";
+            _out << nl << contextParamName << " : Ice.Context";
+            _out << nl << "    The request context for the invocation.";
         }
         if (mode == DocDispatch || mode == DocAsyncDispatch)
         {
             const string currentParamName = getEscapedParamName(op, "current");
-            _out << nl << currentParamName << " -- The Current object for the invocation.";
+            _out << nl << currentParamName << " : Ice.Current";
+            _out << nl << "    The Current object for the invocation.";
         }
     }
-    else if (mode == DocAsyncEnd)
-    {
-        _out << nl << "Arguments:" << nl << "_r - The asynchronous result object for the invocation.";
-    }
 
     //
     // Emit return value(s).
     //
+    bool hasReturnValue = false;
     if (mode == DocAsync || mode == DocAsyncDispatch)
     {
-        _out << nl << "Returns: A future object for the invocation.";
-    }
-    if (mode == DocAsyncBegin)
-    {
-        _out << nl << "Returns: An asynchronous result object for the invocation.";
+        hasReturnValue = true;
+        if (!comment.description.empty() || needArgs)
+        {
+            _out << nl;
+        }
+        _out << nl << "Returns";
+        _out << nl << "-------";
+        _out << nl << "Ice.Future";
+        _out << nl << "    A future object that is completed with the result of "
+             << (mode == DocAsync ? "the invocation." : "the dispatch.");
```

Comment:
This does not look correct. Apparently DocSync and DocAsync are about invocations, while DocDispatch and DocAsyncDispatch are about dispatches (these names aren't consistent).

Reply:
Here we handle returning a future from Async versions, so is either DocAsycn or DocAsyncDispatch.

DocAsync -> "the invocation"
DocAsyncDispatch "the dispatch"

looks correct to me, but I might be missing your point.

Reply:
You're right, I missed the `if (mode == DocAsync || mode == DocAsyncDispatch)`.